lib/cisco_node_utils/ace.rb

```diff
@@ -0,0 +1,225 @@
+
+module Cisco
+  # Ace - node utility class for Ace Configuration
+  class Ace < NodeUtil
+    attr_reader :acl_name, :seqno, :afi
+    # acl_name is name of acl
+    # seqno is sequence number of ace
+    # afi is either v4 or v6
+    def initialize(acl_name, seqno, afi)
+      fail TypeError unless
+        acl_name.is_a?(String) && seqno.is_a?(Integer) && afi.is_a?(String)
+      fail ArgumentError 'we expect ip or ipv6' unless
+        afi == 'ip' || afi == 'ipv6'
+      @acl_name = acl_name
+      @afi = afi
+      @seqno = seqno
+      @set_args = @get_args = { acl_name: @acl_name, seqno: @seqno, afi: @afi,
+                                src_port: '', dst_port: '',
+                                option_format: '' }
+      @regexp_str = '(?<seqno>\d+) (?<action>\S+)'\
+                 ' *(?<proto>\d+|\S+)'\
+                 ' *(?<src_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)*'\
+                 ' *(?<src_port>eq \S+|neq \S+|lt \S+|''gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<dst_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)'\
+                 ' *(?<dst_port>eq \S+|neq \S+|lt \S+|gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<option>[a-zA-Z0-9\-\/ ]*)*'
+    end
+
+    # Create a hash of all aces under give acl_name.
+    def self.aces
+      regexp_str = '(?<seqno>\d+) (?<action>\S+)'\
+                 ' *(?<proto>\d+|\S+)'\
+                 ' *(?<src_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)*'\
+                 ' *(?<src_port>eq \S+|neq \S+|lt \S+|''gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<dst_addr>any|host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+|'\
+                 'addrgroup \S+)'\
+                 ' *(?<dst_port>eq \S+|neq \S+|lt \S+|gt \S+|range \S+ \S+|'\
+                 'portgroup \S+)?'\
+                 ' *(?<option>[a-zA-Z0-9\-\/ ]*)*'
+      afis = %w(ip ipv6)
+      hash = {}
+      afis.each do |afi|
+        get_args = { afi: afi }
+        instances = config_get('acl', 'all_acl', get_args)
+        next if instances.nil?
+        instances.each do |name|
+          hash[name] = {}
+          get_args[:acl_name] = name
+          aces = config_get('acl', 'all_ace', get_args)
+          next if aces.nil?
+          aces.each do |ace|
+            regexp = Regexp.new(regexp_str)
+            match = regexp.match(ace)
+            item = Ace.new(name, match[1].to_i, afi)
+            item.seqno = match[1].to_i
+            item.action = match[2]
+            item.proto = match[3]
+            item.src_addr = match[4]
+            item.src_port = match[5]
+            item.dst_addr = match[6]
+            item.dst_port = match[7]
+            item.option_format = match[8]
+            hash[name][match[1].to_i] = item
+          end
+        end
+      end
+      hash
+    end
```

Comment:
You don't need to set properties in self.aces, it should look like this instead:

```
  def self.aces
    # Create a hash of AFIs/ACLs/ACEs. We only need the ACE seqno, not the properties.
    # ip access-list bar
    #   10 deny tcp any any
    #   20 permit ip any any
    def self.aces
      afis = %w(ip ipv6)
      hash_final = {}
      afis.each do |afi|
        # Get a list of all acls for this afi
        acls = config_get('acl', 'all_acls', afi: afi)
        #             cli: 'ip access-list foo'
        #      config_get: ['foo']
        next if acls.nil?

        # Walk each acl and get a list of aces for the given acl
        acls.each do |acl_name|
          hash_tmp[acl_name] = {}
          aces = config_get('acl', 'all_aces', afi: afi)
          #             cli: '10 deny tcp any any'
          #      config_get: ['10', '20']
          next if aces.nil?

          aces.each do |seqno|
            hash_tmp[acl_name][seqno] = Ace.new(afi, acl, seqno, false)
          end
        end
        puts "hash_tmp: #{hash_temp}"
      end
      hash_final
    end
  end
```